grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java

```diff
@@ -95,8 +95,16 @@ public static GraqlTraversal createTraversal(
      * @return a semi-optimal traversal plan to execute the given conjunction
      */
     private static List<Fragment> semiOptimalConjunction(ConjunctionQuery query, long maxTraversalAttempts) {
-
-        Set<EquivalentFragmentSet> fragmentSets = Sets.newHashSet(query.getEquivalentFragmentSets());
+        Plan initialPlan = Plan.base();
+        Set<EquivalentFragmentSet> fragmentSets = query.getEquivalentFragmentSets().stream().filter(fragmentSet -> {
+            if (fragmentSet.fragments().size() == 1) {
+                Fragment fragment = fragmentSet.fragments().iterator().next();
+                if (fragment.hasFixedFragmentCost()) {
+                    return !initialPlan.tryPush(fragment);
+                }
+            }
+            return true;
+        }).collect(Collectors.toSet());
```

Comment:
We should have a comment to explain what this does, or extract it to a method with a good name.